### PR TITLE
[Ck_tile] smoothquant

### PR DESCRIPTION
**Pull request by rocking5566**

### CODE REVIEW

File: example/ck_tile/12_smoothquant/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+set(TILE_SMOOTHQUANT_COMPILE_OPTIONS)
+
+# NOTE: we turn off undefined-func-template to let source compile without explicit declare function specializations
+list(APPEND TILE_SMOOTHQUANT_COMPILE_OPTIONS -Wno-undefined-func-template -Wno-float-equal)
+
+target_compile_options(${TILE_SMOOTHQUANT} PRIVATE ${TILE_SMOOTHQUANT_COMPILE_OPTIONS})
+
+set(EXAMPLE_SMOOTHQUANT "tile_example_smoothquant")
+add_executable(${EXAMPLE_SMOOTHQUANT} EXCLUDE_FROM_ALL example_smoothquant.cpp)
```
Comment (poyenc) [inline example/ck_tile/12_smoothquant/CMakeLists.txt]: please consider extract common commands out as function

File: example/ck_tile/12_smoothquant/instances/smoothquant_bf16_n256_instance.cpp [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2024, Advanced Micro Devices, Inc. All rights reserved.
+
+#include "smoothquant_instance_common.hpp"
+
+// clang-format off
+//                                                   rm rn tm  tn   vn   pd    2p
```
Comment (poyenc) [inline example/ck_tile/12_smoothquant/instances/smoothquant_bf16_n256_instance.cpp]: Are the titles not aligned?

File: example/ck_tile/12_smoothquant/script/perf_test.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
+
+# run from top of ck folder
+EXE=build/bin/tile_smoothquant
```
Comment (poyenc) [inline example/ck_tile/12_smoothquant/script/perf_test.sh]: Not everyone name the build directory exactly as **build**, I suggest to use `find` command to find available executable to run (under current folder)
```bash
EXE="$(find . -name tile_smoothquant -type f | head -n 1)"
```

File: example/ck_tile/12_smoothquant/instances/smoothquant_fwd_api.cpp [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    float r = -1;
+    if(t.data_type.compare("fp16") == 0)
+    {
+        return smoothquant_dispatch<ck_tile::fp16_t>(t, a, s);
+    }
+    else if(t.data_type.compare("bf16") == 0)
+    {
+        return smoothquant_dispatch<ck_tile::bf16_t>(t, a, s);
+    }
+    if(r < 0)
```
Comment (poyenc) [inline example/ck_tile/12_smoothquant/instances/smoothquant_fwd_api.cpp]: you can simply use `else` clause here and remove the `r` variable

File: example/ck_tile/12_smoothquant/instances/smoothquant_fwd_api.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                                   Vector_N_,
+                                   kPadN_,
+                                   kTwoPass_>;
+
+template <typename data_type>
+float smoothquant_dispatch(smoothquant_traits /*t*/,
+                           smoothquant_args a,
+                           const ck_tile::stream_config& s)
+{
+#if 1
```
Comment (poyenc) [inline example/ck_tile/12_smoothquant/instances/smoothquant_fwd_api.cpp]: please use a macro like `DEBUG_SMOOTHQUANT` to turn on/off debug code

File: include/ck_tile/ops/smoothquant/pipeline/smoothquant_pipeline_problem.hpp [hunk 6]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (poyenc) [inline include/ck_tile/ops/smoothquant/pipeline/smoothquant_pipeline_problem.hpp]: update the year

File: include/ck_tile/ops/smoothquant/pipeline/smoothquant_pipeline_default_policy.hpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            tile_distribution_encoding<
+                sequence<>,
+                tuple<sequence<S::Repeat_M, S::WarpPerBlock_M, S::ThreadPerWarp_M, S::Vector_M>,
+                      sequence<S::Repeat_N, S::WarpPerBlock_N, S::ThreadPerWarp_N, S::Vector_N>>,
+                tuple<sequence<1, 2>, sequence<1, 2>>,
+                tuple<sequence<1, 1>, sequence<2, 2>>,
+                sequence<1, 1, 2, 2>,
+                sequence<0, 3, 0, 3>>{});
+    }
+    template <typename Problem>
```
Comment (poyenc) [inline include/ck_tile/ops/smoothquant/pipeline/smoothquant_pipeline_default_policy.hpp:30]: missing blank line between function definitions

File: include/ck_tile/ops/smoothquant/kernel/smoothquant_kernel.hpp [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    };
+    using Hargs = SmoothquantHostArgs;
+
+    CK_TILE_HOST static constexpr Kargs MakeKargs(const Hargs& hargs)
+    {
+        return Kargs{
+            hargs.p_x, hargs.p_xscale, hargs.p_yscale, hargs.p_qy, hargs.m, hargs.n, hargs.stride};
+    }
+
+    CK_TILE_HOST static constexpr auto GridSize(const Hargs& hargs)
```
Comment (poyenc) [inline include/ck_tile/ops/smoothquant/kernel/smoothquant_kernel.hpp:71]: explicitly cast return value to `dim3`

File: include/ck_tile/ops/smoothquant/kernel/smoothquant_kernel.hpp [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    {
+        return (hargs.m + Block_M - 1) / Block_M;
```
Comment (poyenc) [inline include/ck_tile/ops/smoothquant/kernel/smoothquant_kernel.hpp]: please explicitly cast return value to `dim3`

File: example/ck_tile/12_smoothquant/instances/smoothquant_bf16_n256_instance.cpp (hunk 2 above)
Comment (rocking5566) [reply example/ck_tile/12_smoothquant/instances/smoothquant_bf16_n256_instance.cpp]: cb53ad1dac7c32bdee8ff9938fd0f55c3b61d2d2

File: example/ck_tile/12_smoothquant/script/perf_test.sh (hunk 3 above)
Comment (rocking5566) [reply example/ck_tile/12_smoothquant/script/perf_test.sh]: e9dc1a4af35c0da075649b85a11cebec0b18b6b8

File: example/ck_tile/12_smoothquant/instances/smoothquant_fwd_api.cpp (hunk 4 above)
Comment (rocking5566) [reply example/ck_tile/12_smoothquant/instances/smoothquant_fwd_api.cpp]: 6912b1b8fb990b9160e24370ec18441c2ff1dbd1

File: example/ck_tile/12_smoothquant/instances/smoothquant_fwd_api.cpp (hunk 5 above)
Comment (rocking5566) [reply example/ck_tile/12_smoothquant/instances/smoothquant_fwd_api.cpp]: 9531b2226d4fe39d72b6903c0b61f87c62dd5bfe

File: include/ck_tile/ops/smoothquant/pipeline/smoothquant_pipeline_problem.hpp (hunk 6 above)
Comment (rocking5566) [reply include/ck_tile/ops/smoothquant/pipeline/smoothquant_pipeline_problem.hpp]: 16a2a5b8a52ff76849f2117f4ec72f6722e79e2f

File: include/ck_tile/ops/smoothquant/pipeline/smoothquant_pipeline_default_policy.hpp (hunk 7 above)
Comment (rocking5566) [reply include/ck_tile/ops/smoothquant/pipeline/smoothquant_pipeline_default_policy.hpp:30]: da7980141641ae7a23c5cc5daac734793fda58ca

File: include/ck_tile/ops/smoothquant/kernel/smoothquant_kernel.hpp (hunk 9 above)
Comment (rocking5566) [reply include/ck_tile/ops/smoothquant/kernel/smoothquant_kernel.hpp]: c9ceb118ba10d7f32af4d8f949da8fde169274b9

File: example/ck_tile/12_smoothquant/example_smoothquant.cpp [hunk 10]
```diff
@@ -0,0 +1,237 @@
+#include "ck_tile/host.hpp"
+#include "ck_tile/core.hpp"
+#include "ck_tile/host/kernel_launch.hpp"
+#include "ck_tile/ops/smoothquant.hpp"
+#include <cstring>
+
+// different threshold for different dtype
+template <typename DataType>
+auto get_elimit()
```
Comment (poyenc) [inline example/ck_tile/12_smoothquant/example_smoothquant.cpp:9]: please consider extract common function declarations into another header.

File: example/ck_tile/12_smoothquant/CMakeLists.txt (hunk 1 above)
Comment (rocking5566) [reply example/ck_tile/12_smoothquant/CMakeLists.txt]: e0068dd065d52764542cb6ca6b59a3efbb809fe2

File: example/ck_tile/12_smoothquant/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    endforeach()
+
+    target_sources(${TARGET_NAME} PRIVATE ${INSTANCE_SRCS})
+
+    set(COMPILE_OPTIONS)
+    # NOTE: we turn off undefined-func-template to let source compile without explicit declare function specializations
+    list(APPEND COMPILE_OPTIONS -Wno-undefined-func-template -Wno-float-equal)
+
+    target_compile_options(${TARGET_NAME} PRIVATE ${COMPILE_OPTIONS})
+endfunction(add_smoothquant_example TARGET_NAME MAIN_SRC COMPILE_OPTIONS INSTANCE_SRCS)
```
Comment (poyenc) [inline example/ck_tile/12_smoothquant/CMakeLists.txt]: forgot to update here

**Review by poyenc — APPROVED** [review]

**Pull request merged**